### PR TITLE
[MBO-790][MBO-577][MBO-816] Execute module actions in a Queue

### PR DESCRIPTION
**Pull request by sowbiba**

### CODE REVIEW

File: src/Module/Action/AbstractAction.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    protected $status;
+
+    public function __construct(string $status)
+    {
+        $this->status = $status;
+    }
+
+    public function execute(): bool
+    {
+        throw new \Exception('Method execute must be implented. You\'re using the abstract action.');
```
Comment (sowbiba) [inline src/Module/Action/AbstractAction.php:35]:
```suggestion
        throw new \Exception('Method execute must be implemented. You\'re using the abstract action.');
```

File: src/Module/Action/ActionRetriever.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License version 3.0
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\Module\Mbo\Module\Action;
+
+use Db;
+use Ramsey\Uuid\Uuid;
```
Comment (sowbiba) [inline src/Module/Action/ActionRetriever.php:25]:
```suggestion
```

File: src/Module/Action/ActionRetriever.php [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        $dateAdd = (new \DateTime())->format('Y-m-d H:i:s');
+        $sql = 'INSERT INTO `' . _DB_PREFIX_ . 'mbo_action_queue`(`action_uuid`,`action`,`module`,`parameters`,`status`,`date_add`) VALUES ';
+        $sql .= sprintf(
+            "('%s', '%s', '%s', %s, '%s', '%s');",
+            $action->getActionUuid(),
+            $action->getActionName(),
+            $action->getModuleName(),
+            !empty($action->getParameters()) ? "'" . json_encode($action->getParameters()) . "'" : 'NULL',
+            $action->getStatus(),
+            $dateAdd
```
Comment (sowbiba) [inline src/Module/Action/ActionRetriever.php:60]: Use NOW ?

File: src/Module/Action/InstallAction.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        $this->actionUuid = $actionUuid;
+        $this->moduleName = $moduleName;
+        $this->source = $source;
+
+        parent::__construct($status);
+    }
+
+    public function execute(): bool
+    {
+        // Notify Distribution API that install action is being process
```
Comment (sowbiba) [inline src/Module/Action/InstallAction.php:72]:
```suggestion
        // Notify Distribution API that install action is in progress
```

File: src/Module/Action/InstallAction.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        return $this->moduleManager;
+    }
+
+    public function getActionUuid(): string
+    {
+        return $this->actionUuid;
+    }
+
+    public function getModuleName(): string
```
Comment (sowbiba) [inline src/Module/Action/InstallAction.php:95]: These getters can be in the abstract

**Review by intraordinaire — APPROVED** [review]
Wonderful !

**Review by intraordinaire — APPROVED** [review]

**Pull request closed without merging**